appIcons.js

```diff
@@ -938,6 +941,73 @@ const MyAppIcon = new Lang.Class({
         cr.$dispose();
     },
 
+    _notificationBadge: function() {
+        this._notificationBadgeLabel = new St.Label();
+        this._notificationBadgeBin = new St.Bin({
+            child: this._notificationBadgeLabel,
+            x_align: St.Align.END, y_align: St.Align.START,
+            x_expand: true, y_expand: true
+        });
+        this._notificationBadgeLabel.add_style_class_name('notification-badge');
+        this._notificationBadgeCount = 0;
+        this._notificationBadgeBin.hide();
+
+        this._iconContainer.add_child(this._notificationBadgeBin);
+        this._iconContainer.connect('allocation-changed', Lang.bind(this, this.updateNotificationBadge));
+    },
+
+    updateNotificationBadge: function() {
+        let scaleFactor = St.ThemeContext.get_for_stage(global.stage).scale_factor;
+        let [minWidth, natWidth] = this._iconContainer.get_preferred_width(-1);
+        let logicalNatWidth = natWidth / scaleFactor;
+        let font_size = Math.max(10, Math.round(logicalNatWidth / 5));
+        let margin_left = Math.round(logicalNatWidth / 4);
+
+        this._notificationBadgeLabel.set_style(
+           'font-size: ' + font_size + 'px;' +
+           'margin-left: ' + margin_left + 'px;'
+        );
+
+        this._notificationBadgeBin.width = Math.round(logicalNatWidth - margin_left);
+        this._notificationBadgeLabel.clutter_text.ellipsize = Pango.EllipsizeMode.MIDDLE;
+    },
+
+    _notificationBadgeCountToText: function(count) {
+        if (count <= 9999) {
+            return count.toString();
+        } else if (count < 10**5) {
+            let thousands = count / 10**3;
```

Comment:
Quick observation: the 10**x syntax is not supported on gjs version of my system (Debian Stretch, Gnome 3.22). For retrocompatibily -- so far I have been supporting 3.18--3.24 with the same codebase -- better stick to 1ex format form the moment.